### ISSUE TITLE
Intermittent Workload CLI errors: 'Number of entries expected in End of Central Directory...'

### ISSUE THREAD

**Issue by Redth** [dotnet/sdk#20849]
We are seeing increasing reports of this particular error when running `dotnet workload *` commands from the CLI:

```
Workload installation failed: Number of entries expected in End Of Central Directory does not correspond to number of entries in Central Directory.
```

I've observed this using the NuGet Client API libraries myself with the `CopyNupkgFileToAsync(..)` method: https://github.com/Redth/dotnet-maui-check/blob/main/MauiCheck/AcquirePackagesCommand.cs#L196

It seems to be what's happening is at some point the NuGet client is downloading to the cache and storing a corrupt .nupkg (zip) file.  The problem is, once this is in the cache, it will not try and correct it by downloading it again.

This is really more of a NuGet client issue (I'll file an issue there and link back here) but it does ultimately affect workload commands.  Perhaps there could be something done to add retry logic or wipe the failed cache item when this is encountered to improve success?

**Comment by Redth** [comment]
//cc @wli3  @dsplaisted  @nkolev92

**Comment by nkolev92** [comment]
It might be helpful to have more details as to how you root caused the problem. 

In particular, I'm curious how you determined, the following is happening: 

> It seems to be what's happening is at some point the NuGet client is downloading to the cache and storing a corrupt .nupkg (zip) file.

cc @zivkan

**Comment by Redth** [comment]
@nkolev92 it's slightly anecdotal, but I had my machine in a state where every time i hit the same exception on the same package id when calling `CopyNupkgFileToAsync(..)` in my code.  As soon as I cleared my nuget cache and tried again, it succeeded.

**Comment by Redth** [comment]
Also the 'fix' generally speaking has been telling users to clear their nuget cache which generally resolves the issue, or at least causes a different package to be the point of failure.

**Comment by nkolev92** [comment]
I see. 

Adding @dtivel and @heng-liu who are the signing experts on our team. 

We have seen this manifestation in the past where certain packages have been tampered with afaik.

**Comment by dtivel** [comment]
I don't see anything specific to signing.  If the package is malformed/corrupted/tampered, it can lead to this error.

https://source.dot.net/#System.IO.Compression/System/IO/Compression/ZipArchive.cs,504

**Comment by Redth** [comment]
I just observed this again. 

I was able to clear just the nuget http-cache (`dotnet nuget locals http-cache --clear`) and try again and succeed.

Hopefully that helps isolate the point at which there is a corrupt file introduced.

**Comment by zivkan** [comment]
I had a bit of a look at our code, but I believe that if any exceptions happen while downloading the nupkg, that it will not cause the http-cache to contain the partial download. Therefore, I believe that from an application point of view, if the download completes, it got the bytes exactly as .NET's HttpClient gave it to us (therefore neither the network driver nor operating system's TCP stack detected corrupted packets).

I know that NuGet as a product supports ignoring the http cache so everything is downloaded again. It will take a bit of investigation to figure out how this concept is represented in our APIs, and whether or not it's easy to use the http cache for first attempts, and re-download one specific http-cache entry for retries.

An idea to have NuGet verify the package while downloading, for example taking a checksum within NuGet.Protocol, has two problems:
1. It requires a change to the protocol, which means both servers and clients have to implement it. This makes it much, much slower to work end-to-end.
2. It's going to have a performance impact on restore, NuGet's most performance critical scenario.

I still haven't seen any other customers report this issue. Does the Xamarin/MAUI/dotnet cli teams have issues you can link/share with us, so we can try to understand the extent of the issue and help prioritize against other work?

Also, is Azure DevOps the only source tested for workload downloads?  In other words, do we have any evidence it's a NuGet Client issue rather than an Azure DevOps issue?

**Comment by marcpopMSFT** [comment]
@zivkan are you tracking this on the nuget side with NuGet/Home#11226?  Anything you need from us on this?

**Comment by zivkan** [comment]
If you have any data on the number of customers impacted, the frequency of which it happens, or anything similar, that can help us prioritize this work over others, that will help. Otherwise we can close this issue here and track it in NuGet/Home.

**Comment by Redth** [comment]
I don't have a lot of data or logs to share unfortunately other than we've seen probably a dozen or so internal users hit this (seems potentially weighted more towards non-US based), we've seen it out in external users too. I've probably seen it 10 times myself alone over the last month or two - @mattleibow i believe has seen this a number of times too. 

However I'm not sure it has been observed or not from a non AzDo feed. It could very well be an AzDo specific issue. 

If that may be the case, how do we proceed?

**Comment by rolfbjarne** [comment]
I run into this too once in a while (I live in Spain, with a 1gbps internet connection), but in my case I haven't been able to reproduce the error again after it happens (I do whatever I did again, and it succeeds). So is there any way to configure my system to get better diagnostics when it happens?

**Comment by zivkan** [comment]
I wrote this little program that uses the Nuget Client SDK to download a nupkg, and uses EventListener to listen to HttpClient's internal debugging events, and writes the HTTP response headers (and some other events with the same event and method name) to the console.

<details>
<summary>Source code</summary>
<p>

```cs
using System.Diagnostics.Tracing;
using NuGet.Common;
using NuGet.Protocol;
using NuGet.Protocol.Core.Types;
using NuGet.Versioning;

var listener = new Listener();

var source = Repository.Factory.GetCoreV3("https://pkgs.dev.azure.com/dnceng/public/_packaging/nuget-build/nuget/v3/index.json");
var cache = new SourceCacheContext()
{
    NoCache = true
};
var resource = await source.GetResourceAsync<FindPackageByIdResource>();

string packageId = "NuGet.Protocol";
NuGetVersion version = NuGetVersion.Parse("6.0.0-rc.258");

using (var memoryStream = new MemoryStream())
{
    await resource.CopyNupkgToStreamAsync(
        packageId,
        version,
        memoryStream,
        cache,
        NullLogger.Instance,
        CancellationToken.None);

    Console.WriteLine("Downloaded nupkg, it is {0} bytes", memoryStream.Length);

    var zipArchive = new ZipArchive(memoryStream);
    var files = zipArchive.Entries.Count;
    Console.WriteLine("nupkg contains {0} files", files);
}

class Listener : EventListener
{
    static readonly IReadOnlyDictionary<string, IReadOnlyCollection<string>> subscriptions =
        new Dictionary<string, IReadOnlyCollection<string>>()
        {
            ["Private.InternalDiagnostics.System.Net.Http"] = new HashSet<string>()
            {
                "HandlerMessage"
            }
        };

    protected override void OnEventSourceCreated(EventSource eventSource)
    {
        if (subscriptions.ContainsKey(eventSource.Name))
        {
            EnableEvents(eventSource, EventLevel.LogAlways);
        }
    }

    protected override void OnEventWritten(EventWrittenEventArgs eventData)
    {
        var subscribedEvents = subscriptions[eventData.EventSource.Name];
        if (subscribedEvents.Contains(eventData.EventName))
        {
            if (eventData.EventSource.Name == "Private.InternalDiagnostics.System.Net.Http"
                && eventData.EventName == "HandlerMessage")
            {
                for (int i =0; i < eventData.PayloadNames?.Count; i++)
                {
                    if (eventData.PayloadNames[i] == "memberName")
                    {
                        var memberName = eventData.Payload[i].ToString();
                        if (memberName != "SendAsyncCore")
                        {
                            return;
                        }
                    }
                }
            }

        Console.WriteLine($"***** {eventData.EventSource.Name}::{eventData.EventName}");
        for (int i = 0; i < eventData.Payload?.Count; i++)
        {
            try{
            var name = eventData.PayloadNames[i];
            var value = eventData.Payload[i].ToString();
            Console.WriteLine($"  {name}: {value}");
            }
            catch
            {

            }
        }
        Console.WriteLine("*****");
        }
    }
}
```

</p>
</details>


We can change the source URL and package ID and version to one that one of you have experienced being corrupt. Run it in a loop, with a 1 minute delays between attempts to avoid hammering AzDO's servers, and when the exit code is non-zero, check the output to see if we can spot anything interesting, like if the memory stream length is different to the http response's Content-Length value.

I also see that AzDO has a x-ms-blob-content-md5, so we can adapt this little program to also output the md5, to see if that helps us detect corruption better. NuGet.org doesn't have the same header, so it's not usable in the general sense, but if we want to validate whether AzDO itself has a problem, it's fine.

There are also headers like x-ms-request-id, and ActivityId with guids, that I'm confident are used for distributed tracing, so we can contact the relevant teams and tell them that request id XXXX returned a corrupt file, which might help them investigate.

As part of writing this little program, I found the way to have NuGet.Protocol not use its http-cache is by doing:

```cs
var cache = new SourceCacheContext()
{
    NoCache = true
};
```

and pass that cache context to the download API. Obviously this is unkind to customer's bandwidth and the nuget source's bandwidth, so I wouldn't recommend it as a first attempt, but if you want to catch ZipArchive's exception and then use this as a retry, it's probably suitable.

The NuGet team's own tooling (Visual Studio, nuget.exe, dotnet restore) only uses files in the http-cache for 30 minutes since it was downloaded. After 30 minutes, it will redownload (I'm not 100% sure, but I believe this includes nupkgs, despite the fact that nupkgs are supposed to be immutable. The cache appears designed for json, so the 30 minutes ensures that new package versions are eventually discovered). I don't know if that's settings we set on the SourceCacheContext object when we create it in our own product code, or if it's the default when you use the cache context's default constructor without setting the max age property.

Anyway, if you let me know the feed and package ID of a package that has in the past been a corrupt download, I'll try my program in a loop and see if I can detect any corrupt downloads, and if so, I'll get in contact with the relevant teams to see if we can help them investigate.

If we find that the problem is not limited to one source, then I can get in touch with the System.Net.Http team and see if they want to investigate, to see if we can determine the root cause, or if it's just the internet being unreliable and if it's something that NuGet must handle itself.

**Comment by rolfbjarne** [comment]
Here's a verbose log from a "dotnet workload install" that failed: https://gist.github.com/rolfbjarne/abc210f789a356cfbc3b9133bca9fe15

One interesting part is that a nupkg failed to download (it timed out), then apparently it tried again and that succeeded (supposedly?): https://gist.github.com/rolfbjarne/abc210f789a356cfbc3b9133bca9fe15#file-gistfile1-txt-L206-L211

**Comment by zivkan** [comment]
Thanks, I was wondering what feed and package IDs are affected, so this is helpful. Although if someone could share the service index URL (that you put in nuget.config) that would be helpful, as the logs don't contain it, although I think I can guess.

Either on Friday or Monday, I emailed an internal list for System.Net. I got 3 replies, each of which I found very interesting:

1. The first person informed me that TCP header has a 16-bit CRC.
   * We didn't discuss it in the email, but afterwards I was thinking, although TCP technically supports up to a 65k packet, the vast majority of people don't use these "jumbo frames". My own ethernet MTU is 1500, and my VPN network adapter has a 1400 MTU. A 16-bit CRC has much more than 1500 values, meaning it's extremely unlikely that a single bit change caused the TCP CRC to collide with a valid value. Therefore, multiple bit errors must occur, and be unlucky enough to match the TCP CRC for the operating system (TCP/IP stack) to pass the packet onto the client application. I don't know if lower levels like ethernet have error detection as well, if so, it makes the probability of all these errors still resulting in a valid checksum even less likely.
2. The second person works on ASP.NET and said he's never encountered a problem where bytes were transformed in transit, and that corruption was always due to either the server or client truncating the stream. It's an anecdote, not data, but I trust this person, so I think my previous message where I was implicitly suggesting we check file lengths vs HTTP header Content-Length is along a good track.
3. The 3rd person said that TLS has HMAC to validate the encrypted payload. This means that for corrupt data we'd need not only the TCP packets to be corrupt, but still pass CRC, but also for the cryptographic HMAC to be corrupt, yet still valid. The probability of both of these things happening are so close to zero, that given the Azure DevOps feed is over https, I don't think it's worth entertaining the idea of bytes changing over transport.

Given the error message is "Number of entries expected in End Of Central Directory does not correspond to number of entries in Central Directory", this is consistent with the file stream being truncated early.

I've been meaning to create a VM or two in Azure and have a loop downloading the package over and over trying to detect an occurrence of the problem, but I haven't had time. I'll see if I can block a small amount of time next month to dedicate to progressing this investigation.

I provided hints in my previous comment so that non-NuGet products that use the NuGet SDK, such as the `dotnet workload *` commands, and `dotnet-maui-check`, they can detect the zip exception, and then re-try the download, bypassing NuGet's HTTP cache. So, if people working on those tools have more time than me to make improvements, I hope you're unblocked. If we can determine that it's a truncation issue, and that the server is not sending the incorrect Content-Length header, I hope we can make NuGet.Protocol handle the retry automatically. If we can't figure it out from the HTTP response, then it's going to be a lot harder to make Nuget relisient, and other products using the NuGet SDK might need code changes. But I'm getting ahead of myself, I need to better understand the problem first.

**Comment by Redth** [comment]
Thanks for your persistence on this, I realize it's not an easy one to reproduce. 

I see you already correctly deduced that Linux is not supported for the workload :)

Can you try macos? I personally have seen it more on macos than windows.

**Comment by mattleibow** [comment]
I just noticed this, but I think it is a result of another issue:

```
[NuGet Manager] Failed to download package 'microsoft.android.runtime.android-arm.31.0.101-preview.11.73' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/microsoft.android.runtime.android-arm/31.0.101-preview.11.73/microsoft.android.runtime.android-arm.31.0.101-preview.11.73.nupkg'.
  The download of 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/microsoft.android.runtime.android-arm/31.0.101-preview.11.73/microsoft.android.runtime.android-arm.31.0.101-preview.11.73.nupkg' timed out because no data was received for 60000ms.
    Exception of type 'System.TimeoutException' was thrown.
  [NuGet Manager] [Info]   GET https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/microsoft.android.runtime.android-arm/31.0.101-preview.11.73/microsoft.android.runtime.android-arm.31.0.101-preview.11.73.nupkg
  [NuGet Manager] [Info]   OK https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/microsoft.android.runtime.android-arm/31.0.101-preview.11.73/microsoft.android.runtime.android-arm.31.0.101-preview.11.73.nupkg 91ms
  Workload installation failed. Rolling back installed packs...
  Uninstalling workload pack Microsoft.Android.Sdk version 31.0.101-preview.11.73.
  Uninstalling workload pack Microsoft.Android.Sdk.BundleTool version 31.0.101-preview.11.73.
  Uninstalling workload pack Microsoft.Android.Ref version 31.0.101-preview.11.73.
  Rolling back pack Microsoft.Android.Sdk installation...
  Rolling back pack Microsoft.Android.Sdk.BundleTool installation...
  Rolling back pack Microsoft.Android.Ref installation...
  Rolling back pack Microsoft.Android.Runtime.android-arm installation...
  Workload installation failed: Number of entries expected in End Of Central Directory does not correspond to number of entries in Central Directory.
D:\a\1\s\src\DotNet\DotNet.csproj(160,5): error MSB3073: The command ""D:\a\1\s\bin/dotnet/dotnet" workload install android-aot --skip-manifest-update --verbosity diag --temp-dir "D:\a\1\s\bin/temp/" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-aspnetcore-ae1a6cbe-3/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-aspnetcore-ae1a6cbe-2/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-aspnetcore-ae1a6cbe-1/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-aspnetcore-ae1a6cbe/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-runtime-4822e3c3-6/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-runtime-4822e3c3-5/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-runtime-4822e3c3-4/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-runtime-4822e3c3-2/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-runtime-4822e3c3-1/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-windowsdesktop-59fea7da-6/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-windowsdesktop-59fea7da-5/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-windowsdesktop-59fea7da-4/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-windowsdesktop-59fea7da-3/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-windowsdesktop-59fea7da-2/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-windowsdesktop-59fea7da-1/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-windowsdesktop-59fea7da/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-emsdk-1ec2e17f-6/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-emsdk-1ec2e17f-5/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-emsdk-1ec2e17f-4/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-emsdk-1ec2e17f-3/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-emsdk-1ec2e17f-2/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-emsdk-1ec2e17f-1/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-emsdk-1ec2e17f/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json" --source "https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" --source "https://aka.ms/dotnet6/nuget/index.json" --source "https://aka.ms/skiasharp-eap/index.json"" exited with code 1.

````

**Comment by akoeplinger** [comment]
The Xamarin Android team just hit this too in their CI on Linux:  https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5404924&view=logs&j=4f348570-eb52-582b-3455-1d0941a5aec9&t=fae33dd4-23e9-51a3-7ed8-abdaa49914c7&l=370

```
  The download of 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/30833ef4-8c6c-4465-9e68-21951b9880d2/nuget/v3/flat2/microsoft.netcore.app.runtime.aot.linux-x64.cross.android-arm/6.0.0/microsoft.netcore.app.runtime.aot.linux-x64.cross.android-arm.6.0.0.nupkg' timed out because no data was received for 60000ms.
    Exception of type 'System.TimeoutException' was thrown.
  [NuGet Manager] [Info]   GET https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/30833ef4-8c6c-4465-9e68-21951b9880d2/nuget/v3/flat2/microsoft.netcore.app.runtime.aot.linux-x64.cross.android-arm/6.0.0/microsoft.netcore.app.runtime.aot.linux-x64.cross.android-arm.6.0.0.nupkg
  [NuGet Manager] [Info]   OK https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/30833ef4-8c6c-4465-9e68-21951b9880d2/nuget/v3/flat2/microsoft.netcore.app.runtime.aot.linux-x64.cross.android-arm/6.0.0/microsoft.netcore.app.runtime.aot.linux-x64.cross.android-arm.6.0.0.nupkg 348ms
  Workload installation failed. Rolling back installed packs...
  Uninstalling workload pack Microsoft.Android.Sdk version 31.0.101-ci.pr.gh6464.77…
  Uninstalling workload pack Microsoft.Android.Sdk.BundleTool version 31.0.101-ci.pr.gh6464.77…
  Uninstalling workload pack Microsoft.Android.Ref version 31.0.101-ci.pr.gh6464.77…
  Uninstalling workload pack Microsoft.Android.Runtime.android-arm version 31.0.101-ci.pr.gh6464.77…
  Uninstalling workload pack Microsoft.Android.Runtime.android-arm64 version 31.0.101-ci.pr.gh6464.77…
  Uninstalling workload pack Microsoft.Android.Runtime.android-x86 version 31.0.101-ci.pr.gh6464.77…
  Uninstalling workload pack Microsoft.Android.Runtime.android-x64 version 31.0.101-ci.pr.gh6464.77…
  Uninstalling workload pack Microsoft.Android.Templates version 31.0.101-ci.pr.gh6464.77…
  Uninstalling workload pack Microsoft.NETCore.App.Runtime.Mono.android-arm version 6.0.0…
  Uninstalling workload pack Microsoft.NETCore.App.Runtime.Mono.android-arm64 version 6.0.0…
  Uninstalling workload pack Microsoft.NETCore.App.Runtime.Mono.android-x64 version 6.0.0…
  Uninstalling workload pack Microsoft.NETCore.App.Runtime.Mono.android-x86 version 6.0.0…
  Uninstalling workload pack Microsoft.NET.Runtime.MonoAOTCompiler.Task version 6.0.0…
  Uninstalling workload pack Microsoft.NET.Runtime.MonoTargets.Sdk version 6.0.0…
  Uninstalling workload pack Microsoft.NETCore.App.Runtime.AOT.Cross.android-x86 version 6.0.0…
  Uninstalling workload pack Microsoft.NETCore.App.Runtime.AOT.Cross.android-x64 version 6.0.0…
  Rolling back pack Microsoft.Android.Sdk installation...
  Rolling back pack Microsoft.Android.Sdk.BundleTool installation...
  Rolling back pack Microsoft.Android.Ref installation...
  Rolling back pack Microsoft.Android.Runtime.android-arm installation...
  Rolling back pack Microsoft.Android.Runtime.android-arm64 installation...
  Rolling back pack Microsoft.Android.Runtime.android-x86 installation...
  Rolling back pack Microsoft.Android.Runtime.android-x64 installation...
  Rolling back pack Microsoft.Android.Templates installation...
  Rolling back pack Microsoft.NETCore.App.Runtime.Mono.android-arm installation...
  Rolling back pack Microsoft.NETCore.App.Runtime.Mono.android-arm64 installation...
  Rolling back pack Microsoft.NETCore.App.Runtime.Mono.android-x64 installation...
  Rolling back pack Microsoft.NETCore.App.Runtime.Mono.android-x86 installation...
  Rolling back pack Microsoft.NET.Runtime.MonoAOTCompiler.Task installation...
  Rolling back pack Microsoft.NET.Runtime.MonoTargets.Sdk installation...
  Rolling back pack Microsoft.NETCore.App.Runtime.AOT.Cross.android-x86 installation...
  Rolling back pack Microsoft.NETCore.App.Runtime.AOT.Cross.android-x64 installation...
  Rolling back pack Microsoft.NETCore.App.Runtime.AOT.Cross.android-arm installation...
  Workload installation failed: Number of entries expected in End Of Central Directory does not correspond to number of entries in Central Directory.
```

According to the logs this was using a `Standard_D4s_v3` Azure VM using Ubuntu 20.04, but not sure in which region (but it was a 1ES VMSS pool so probably not that far away).

It looks to me like the TimeoutException is a common thing in the last few reports. I wonder if NuGet somehow tries to read the timed-out and probably partially downloaded .nupkg?

**Comment by zivkan** [comment]
I don't know how exactly the dotnet cli is using NuGet's APIs, but as long as it's not [telling NuGet to skip the HTTP cache, NuGet will create a http cache file](https://github.com/NuGet/NuGet.Client/blob/31c1b9f8e3471b93dc30538f84088acbccb170a2/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSource.cs#L157-L163). When it does this, it uses [an overload of `new FileStream`, passing in `FileMode.Create`](https://github.com/NuGet/NuGet.Client/blob/31c1b9f8e3471b93dc30538f84088acbccb170a2/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpCacheUtility.cs#L69-L74), which will [truncate the file if it already exists](https://docs.microsoft.com/en-us/dotnet/api/system.io.filemode?view=net-5.0#fields).

If the dotnet workload command is telling NuGet to use `DirectDownload` (skip the HTTP cache), [there's another codepath](https://github.com/NuGet/NuGet.Client/blob/31c1b9f8e3471b93dc30538f84088acbccb170a2/src/NuGet.Core/NuGet.Protocol/Utility/FindPackagesByIdNupkgDownloader.cs#L169-L190) that has too many layers of abstraction for me to understand at 6am. I think there's a bug/design flaw where the retry will just concatenate the new download from the end of the incomplete download. When the download completes successfuly, then all the file offsets written in the zip file will be wrong, and I wouldn't be surprised if this causes the central directory exceptions.

**Comment by zivkan** [comment]
Ok, [the dotnet cli is using `DirectDownload = true`](https://github.com/dotnet/sdk/blob/22c4860dcb2cf6b123dd641cc4a87a50380759d5/src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs#L63-L68).

If the SDK team is willing to turn this off, then retries should no longer cause corrupted zip files, but it will mean there will be an extra copy of the nupkg in nuget's http-cache which will only be deleted if the customer explicitly runs `dotnet nuget locals http-cache --clear`.

When I come back to this task later this sprint, I'll see if I can get retries to seek to the start of file.

**Comment by philippjbauer** [comment]
I'm adding my findings here in hope that this is helpful.

My internet connection (Unlimited LTE provider, I'm very rural) suffers from stalling downloads. They go strong for a while until the download speed tapers off until it stops, then the download is ending in a timeout or corrupted (unfinished) state. Download managers that can retry downloads usually help me, because they can continue from where the download failed.

Another unfortunate problems that the workload installer suffers from is that successful downloads are deleted when they are rolled back. On slow connections, this becomes very frustrating because you might already had 80% of the packages downloaded.

I would suggest to make the `DirectDownload` option that @zivkan mentions accessible through the CLI option. From what I understand this would eliminate both problems I am experiencing. In the meantime, until this gets fixed, I came up with a workaround that let me successfully install the MAUI workload and which should be applicable for any workload one might want to install through the CLI.

I'm pasting it here for reference, I mentioned it in the OP of dotnet/maui#4203

---
---

I found a workaround encountering download issues with the `dotnet workload install` command. This involves some helpful, but unfortunately undocumented, command options I figured out by looking into the source code in the .NET SDK GitHub repo. 😮‍💨

    ⚠️ Attention
    This requires a \*nix shell environment. You can probably get this working in Powershell as well.

    🗒 Note
    The following commands use sources specifically for installation of MAUI Preview 12.
    If you want to use stable workloads the same instructions should apply but without the 
    need for the rollback file or sources.

## Download URLS

First you have to download the URLs to the needed workload nuget packages.

```sh
# Get all URLS
dotnet workload install maui \
    --print-download-link-only \
    --skip-manifest-update \
    --from-rollback-file https://aka.ms/dotnet/maui/preview.12.json \
    --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json \
    --source https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json
```

```sh
# Example Output
Resolving package URLs for workload(s) maui...
	==allPackageLinksJsonOutputStart==
["https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/825db618-e3eb-4426-ba54-b1d6e6c944d8/nuget/v3/flat2/microsoft.netcore.app.runtime.mono.android-arm/6.0.2-mauipre.1.22054.8/microsoft.netcore.app.runtime.mono.android-arm.6.0.2-mauipre.1.22054.8.nupkg",
 ... ]
==allPackageLinksJsonOutputEnd==
```

## Create list of URLs

Then you follow the next steps to create a file with all links, ready to be used with `wget`.

1. Paste package links JSON into a new text file, e.g. `maui.workload.preview-12.txt`.
2. Replace all `,` with newline, remove `[`, `]` and `"`.
3. Create new folder, e.g. `maui.nuget.cache`.
4. Change into new folder.

## Download packages

Use `wget` command  inside of new cache folder to download all packages from text file.

```sh
wget -nv --show-progress -nc -c -i maui.workload.preview-12.txt
```

You can retry as often as needed, use `ctrl-c` to cancel when the download gets stuck and use the command again to continue where the download stopped. `wget` will skip files you already downloaded.

## Install workload from disk

Now use the downloaded files from your local cache to install the workload.

```sh
sudo dotnet workload install maui \
    --from-cache ~/path/to/maui.nuget.cache \
    --skip-manifest-update \
    --from-rollback-file https://aka.ms/dotnet/maui/preview.12.json \
    --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json \
    --source https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json
```

## Use the same sources to restore a project

```sh
dotnet restore \
    --source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json \
    --source https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json
```

**Comment by zivkan** [comment]
@philippjbauer I'm sorry to hear of the bad experieces you've had with NuGet failing to download files.

By any chance do you have advice on how someone could emulate connections like yours?  I've tried [Clumsy](https://jagt.github.io/clumsy/), but haven't found good settings to make large downloads fail in the way you described. I've thought about trying to implement a "download manager" in C#, to allow NuGet to resume broken downloads, rather than restart. But it's difficult to test, and we just don't have telemetry to give us an indication to how many customers are experiencing such problems in order to prioritize this kind of improvement.

I plan in the next two weeks to resume efforts in getting NuGet to at least reset the filestream, so when a download completes, it's no longer corrupted. But it will use extra bandwidth, because it won't resume from where it stopped.

**Comment by philippjbauer** [comment]
I agree, this is a tricky problem to investigate, especially if telemetry is missing.

Unfortunately I don't know of a way to simulate a connection problem like mine. And I don't understand fully why my network is behaving the way it is either. I got a strong wireless connection to the cell tower 🤷‍♂️

But I can offer to try out your solution when it becomes available.

Thank you for your effort!

**Comment by zivkan** [comment]
NuGet has a code fix for this. Our next insertion into the .NET SDK (next week) will add it into .NET 7 and .NET 6.0.3xx daily builds.
